src/core/net/ip4_headers.hpp

```diff
@@ -0,0 +1,492 @@
+    // |      TTL      |    Protocol   |        Header Checksum        |
+    // |                       Source IP Address                       |
+    // |                         Dest IP Address                       |
+    // +-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+
+
+    static constexpr uint8_t  kVersion4           = 0x40;   // Use with `mVersIhl`
+    static constexpr uint8_t  kVersionMask        = 0xf0;   // Use with `mVersIhl`
+    static constexpr uint8_t  kIhlMask            = 0x0f;   // Use with `mVersIhl`
+    static constexpr uint8_t  kDscpOffset         = 2;      // Use with `mDscpEcn`
+    static constexpr uint16_t kDscpMask           = 0xfc;   // Use with `mDscpEcn`
+    static constexpr uint8_t  kEcnOffset          = 0;      // Use with `mDscpEcn`
+    static constexpr uint8_t  kEcnMask            = 0x03;   // Use with `mDscpEcn`
+    static constexpr uint16_t kFlagsMask          = 0xe000; // Use with `mFlagsFragmentOffset`
+    static constexpr uint16_t kFlagsDf            = 0x4000; // Use with `mFlagsFragmentOffset`
+    static constexpr uint16_t kFlagsMf            = 0x2000; // Use with `mFlagsFragmentOffset`
+    static constexpr uint16_t kFragmentOffsetMask = 0x1fff; // Use with `mFlagsFragmentOffset`
+    static constexpr uint32_t kVersIhlInit        = 0x45;   // Version 4, Header length = 5x8 bytes.
+
+    uint8_t  mVersIhl;
+    uint8_t  mDscpEcn;
+    uint16_t mTotalLength;
+    uint16_t mIdentification;
+    uint16_t mFlagsFargmentOffset;
+    uint8_t  mTtl;
+    uint8_t  mProtocol;
+    uint16_t mHeaderChecksum;
+    Address  mSource;
+    Address  mDestination;
+} OT_TOOL_PACKED_END;
+
+// ICMP(in v4) messages will only be generated / handled by NAT64. So only header defination is required.
+class Icmp
+{
+public:
+    OT_TOOL_PACKED_BEGIN
+    class Header : public Clearable<Header>
+    {
+    public:
+        static constexpr uint16_t kChecksumFieldOffset = 2;
+        // A few ICMP types, only the ICMP types work with NAT64 are listed here.
+        enum Type : uint8_t
+        {
+            kTypeEchoReply              = 0,
+            kTypeDestinationUnreachable = 3,
+            kTypeEchoRequest            = 8,
+            kTypeTimeExceeded           = 11,
+        };
+
+        enum Code : uint8_t
+        {
+            kCodeNone = 0,
+            // Destination Unreachable codes
+            kCodeNetworkUnreachable  = 0,
+            kCodeHostUnreachable     = 1,
+            kCodeProtocolUnreachable = 2,
+            kCodePortUnreachable     = 3,
+            kCodeSourceRouteFailed   = 5,
+            kCodeNetworkUnknown      = 6,
+            kCodeHostUnknown         = 7,
+        };
+
+        /**
+         * This method returns the type of the ICMP message.
+         *
+         * @returns The type field of the ICMP message.
+         *
+         */
+        Type GetType(void) const { return static_cast<Type>(mType); }
```

Comment:
I know we may do this in other places in OT code (which I hope we change later) but I suggest here we just return `uint8_t` as value from `GetType()`.

The reason is that `Type` enum we define does not cover all possible values and we can receive a message with any value for `mType`. I know it technically works correctly (i.e., we can assign any `uint8_t` value to a `enum Type : uint8_t`) but semantically does not fit well.